app/models/discussion.rb

```diff
@@ -152,22 +152,64 @@ def last_versioned_at
     end
   end
 
-  def activity
-    items
-  end
-
   def delayed_destroy
     self.update_attribute(:is_deleted, true)
     self.delay.destroy
   end
 
-  def most_recent_comment
-    comments.order("created_at DESC").first
+
+  def thread_item_created!(item)
+    #update count and last_item_at
+    self.items_count += 1
+    self.last_item_at = item.created_at
+
+    # update first and last sequence ids
+    if self.first_sequence_id == 0
+      self.first_sequence_id = item.sequence_id
+    end
+    self.last_sequence_id = item.sequence_id
+
+    self.save!(validate: false)
   end
 
-  def comment_deleted!
-    refresh_last_comment_at!
-    discussion_readers.each(&:reset_counts!)
+  def thread_item_destroyed!(destroyed_item)
+    self.items_count -= 1
+
+    if destroyed_item.sequence_id == first_sequence_id
+      self.first_sequence_id = sequence_id_or_0(lookup_first_item)
+    end
+
+    if destroyed_item.sequence_id == last_sequence_id
+      last_item = lookup_last_item
+      self.last_sequence_id = sequence_id_or_0(last_item)
+      self.last_item_at = last_item.try(:created_at)
+    end
+
+    self.save!(validate: false)
+
+    discussion_readers.
+      where('last_read_at <= ?', destroyed_item.created_at).
+      each(&:reset_items_count!)
+
+    true
+  end
+
+  def comment_created!(comment)
+    self.comments_count += 1
+    self.last_comment_at = comment.created_at
+
+    self.save!(validate: false)
+  end
+
+  def comment_destroyed!(destroyed_comment)
+    self.comments_count -= 1
+    self.last_comment_at = lookup_last_comment.try(:created_at)
```

Comment:
I'm not totally sure we care about the actual whole record being returned here; you're sorting by `created_at` and then returning comment.created_at, so simpler and faster might be

```
self.comments.maximum(:created_at)
```